### ISSUE TITLE
nan loss not the  hyper-parameter issues,

### ISSUE THREAD

**Issue by rockyzhengwu** [open-mmlab/mmdetection#3511]
**Checklist**

I found this [issue](https://github.com/open-mmlab/mmdetection/issues/3446) but there is no script and data to reproduce ,

**Describe the bug**
train command :
 ```sheel
python tools/train.py ./configs/cascade_rcnn/cascade_rcnn_r50_fpn_1x_coco.py --gpu-ids 1
```
get nan loss:
```sheel
2020-08-08 14:17:58,110 - mmdet - INFO - workflow: [('train', 1)], max: 12 epochs
2020-08-08 14:18:21,885 - mmdet - INFO - Epoch [1][50/58633]	lr: 1.978e-03, eta: 3 days, 20:37:38, time: 0.474, data_time: 0.049, memory: 4109, loss_rpn_cls: 0.4903, loss_rpn_bbox: 0.0879, s0.loss_cls: 1.3479, s0.acc: 85.9355, s0.loss_bbox: 0.0742, s1.loss_cls: 0.6673, s1.acc: 83.2500, s1.loss_bbox: 0.0225, s2.loss_cls: 0.3347, s2.acc: 83.1250, s2.loss_bbox: 0.0032, loss: 3.0280
2020-08-08 14:18:43,873 - mmdet - INFO - Epoch [1][100/58633]	lr: 3.976e-03, eta: 3 days, 17:17:25, time: 0.440, data_time: 0.005, memory: 4109, loss_rpn_cls: 0.2393, loss_rpn_bbox: 0.0847, s0.loss_cls: 0.5997, s0.acc: 92.4023, s0.loss_bbox: 0.1665, s1.loss_cls: 0.1879, s1.acc: 96.1934, s1.loss_bbox: 0.0493, s2.loss_cls: 0.0692, s2.acc: 97.6699, s2.loss_bbox: 0.0073, loss: 1.4038
2020-08-08 14:19:05,444 - mmdet - INFO - Epoch [1][150/58633]	lr: 5.974e-03, eta: 3 days, 15:37:20, time: 0.431, data_time: 0.005, memory: 4109, loss_rpn_cls: 0.1881, loss_rpn_bbox: 0.0714, s0.loss_cls: 0.5252, s0.acc: 93.1387, s0.loss_bbox: 0.1516, s1.loss_cls: 0.1515, s1.acc: 96.6035, s1.loss_bbox: 0.0464, s2.loss_cls: 0.0541, s2.acc: 97.9883, s2.loss_bbox: 0.0078, loss: 1.1960
2020-08-08 14:19:27,544 - mmdet - INFO - Epoch [1][200/58633]	lr: 7.972e-03, eta: 3 days, 15:17:37, time: 0.442, data_time: 0.005, memory: 4109, loss_rpn_cls: 0.2068, loss_rpn_bbox: 0.0727, s0.loss_cls: 0.5826, s0.acc: 92.9688, s0.loss_bbox: 0.1561, s1.loss_cls: 0.1712, s1.acc: 96.5508, s1.loss_bbox: 0.0470, s2.loss_cls: 0.0622, s2.acc: 97.9824, s2.loss_bbox: 0.0064, loss: 1.3051
2020-08-08 14:19:49,435 - mmdet - INFO - Epoch [1][250/58633]	lr: 9.970e-03, eta: 3 days, 14:56:17, time: 0.438, data_time: 0.005, memory: 4109, loss_rpn_cls: 0.3138, loss_rpn_bbox: 0.1092, s0.loss_cls: 0.5817, s0.acc: 93.4922, s0.loss_bbox: 0.1451, s1.loss_cls: 0.1901, s1.acc: 96.7969, s1.loss_bbox: 0.0437, s2.loss_cls: 0.0735, s2.acc: 98.0645, s2.loss_bbox: 0.0071, loss: 1.4642
2020-08-08 14:20:05,850 - mmdet - INFO - Epoch [1][300/58633]	lr: 1.197e-02, eta: 3 days, 11:08:13, time: 0.328, data_time: 0.005, memory: 4109, loss_rpn_cls: nan, loss_rpn_bbox: nan, s0.loss_cls: nan, s0.acc: 41.2221, s0.loss_bbox: nan, s1.loss_cls: nan, s1.acc: 41.3119, s1.loss_bbox: nan, s2.loss_cls: nan, s2.acc: 41.3334, s2.loss_bbox: nan, loss: nan

```

**Reproduction**
I add some code before [this line](https://github.com/open-mmlab/mmdetection/blob/68d860d6f1a4a1fb687f2512038f2cbce7c6b08d/mmdet/models/detectors/two_stage.py#L167) to dump the input when loss is None, the input 
[coco_nan_input.ptn.zip](https://github.com/open-mmlab/mmdetection/files/5044829/coco_nan_input.ptn.zip)

```python
        is_nan = [torch.isnan(l) for l in roi_losses.values()]
        is_nan = [l.cpu().numpy() for l in is_nan]
        if any(is_nan):
            batch = {}
            batch['img'] = img
            batch['gt_labels'] = gt_labels
            batch['gt_bboxes'] = gt_bboxes
            batch['img_metas'] = img_metas
            torch.save(batch, "coco_nan_input.ptn")
            print(losses)
            exit(0)
        return losses
```
then i show the data code is beblow:
``` python
import mmcv
import cv2
import numpy as np

mean=np.array([123.675, 116.28, 103.53])
std=np.array([58.395, 57.12, 57.375])

def vis_imag(image, bbox, labels, filename):
  img = image.data.permute(1,2,0)
  img = img.cpu().numpy()
  img = mmcv.imdenormalize(img, mean, std, False)
  img = np.float32(img)
  labels = labels.cpu().numpy()
  font = cv2.FONT_HERSHEY_SIMPLEX
  for i, box in enumerate(bbox):
    box = box.cpu().numpy()
    img = cv2.rectangle(img, (box[0], box[1]), (box[2], box[3]), (255, 0, 0), 2)
    cv2.putText(img,str(labels[i]),(box[0],box[1]), font, 1 ,(0,255,0),2,cv2.LINE_AA)
  img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
  print("write out: ",filename)
  cv2.imwrite('./%s'%(filename), img)

import torch

def main():
  # args = parse_args()
  batch = torch.load("./coco_nan_input.ptn")
  images = batch['img']
  gt_bboxes = batch['gt_bboxes']
  gt_laels = batch['gt_labels']
  img_metas = batch['img_metas']
  for i in range(len(images)):
    image = images[i].cpu()
    labels = gt_laels[i].cpu()
    bbox = gt_bboxes[i].cpu()
    filename= img_metas[i]['filename']
    vis_imag(image, bbox, labels, filename.split("/")[-1])

if __name__ == "__main__":
  main()
```
there is no problem about the data.
in order to figure out  if this is hyper-parameter issues I write code below:
```python
from mmcv import Config
from mmdet.models import build_detector
import torch

def load_data():
  batch = torch.load('nan_input.ptn')
  return batch

def run():
  config_path = " config_path = "configs/cascade_rcnn/cascade_rcnn_r50_fpn_1x_coco.py"
  cfg = Config.fromfile(config_path)
  batch = load_data()
  model = build_detector(cfg.model, train_cfg=cfg.train_cfg, test_cfg=cfg.test_cfg)
  model.cuda("cuda:1")
  out = model.train_step(batch, optimizer=None)
  print(out)

if __name__ == '__main__':
  run()
```
the output loss is nan , so i think this is not about the hyper-parameter .
```sheel
{'loss': tensor(nan, device='cuda:1', grad_fn=<AddBackward0>), 'log_vars': OrderedDict([('loss_rpn_cls', 0.6916577219963074), ('loss_rpn_bbox', 0.2154594361782074), ('s0.loss_cls', 4.189858913421631), ('s0.acc', 1.46484375), ('s0.loss_bbox', nan), ('s1.loss_cls', 2.201327323913574), ('s1.acc', 0.390625), ('s1.loss_bbox', nan), ('s2.loss_cls', 1.0520288944244385), ('s2.acc', 2.5390625), ('s2.loss_bbox', nan), ('loss', nan)]), 'num_samples': 2}

```

**Environment**
sys.platform: linux
Python: 3.6.6 |Anaconda, Inc.| (default, Jun 28 2018, 17:14:51) [GCC 7.2.0]
CUDA available: True
CUDA_HOME: /usr/local/cuda-10.1
NVCC: Cuda compilation tools, release 10.1, V10.1.105
GPU 0,1: GeForce GTX 1080 Ti
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.5.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2018.0.2 Product Build 20180127 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_INTERNAL_THREADPOOL_IMPL -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF, 

TorchVision: 0.6.0a0+35d732a
OpenCV: 4.2.0
MMCV: 1.0.5
MMDetection: 2.3.0+68d860d
MMDetection Compiler: GCC 7.3
MMDetection CUDA Compiler: 10.1

I install pytorch use conda 

**Error traceback**

**Bug fix**

**Comment by rockyzhengwu** [comment]
maybe I had mixup my data with coco ,  when train coco on one gpu if while get nan loss too. but this may be the hyper-parameter issue .

**Comment by wendalegood** [comment]
hi, rocky, did you solve the problem?

**Comment by rockyzhengwu** [comment]
My data have some problem, I suggest to check the data first
